src/lib/autoLauncher.ts

```diff
@@ -0,0 +1,141 @@
+import * as native from './native'
+
+
+type osAL = typeof macAL | typeof winAL | typeof linAL | typeof nullAL
+
+const nullAL = {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async enable({ appName, appPath, hidden }: AutoLaunchParams): Promise<ChildReturnData> {
+    return { stdout: [], stderr: [] }
+  },
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async disable(appName: string): Promise<ChildReturnData> {
+    return { stdout: [], stderr: [] }
+  },
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async isEnabled(appName: string): Promise<boolean> {
+    return false
+  }
+}
+
+const linAL = {
+
+  async enable({ appName, appPath, hidden }: AutoLaunchParams): Promise<ChildReturnData> {
+    const response: ChildReturnData = { stderr: [], stdout: [] }
+    const hiddenArg = hidden ? ' --hidden' : '';
+
+    const contents = `
+    [Desktop Entry]
+    Type=Application
+    Version=1.0
+    Name=${appName}
+    Comment=${appName}startup script
+    Exec=${appPath}${hiddenArg}
+    StartupNotify=false
+    Terminal=false
+  `
+    await fs.createDir(this.getDirectory()).catch(console.error)
+    await fs.writeFile({ contents, path: this.getFilePath(appName) })
+    response.stdout.push('success')
+    return response
+  },
+  async disable(appName: string): Promise<ChildReturnData> {
+    const response: ChildReturnData = { stderr: [], stdout: [] }
+    await fs.removeFile(this.getFilePath(appName))
+    response.stdout.push("success")
+    return response
+  },
+  async isEnabled(appName: string): Promise<boolean> {
+    try {
+      await fs.readTextFile(this.getFilePath(appName))
+      return true
+    } catch (error) {
+      console.error(error)
+      return false
+    }
+  },
+  getDirectory(): string { return '~/.config/autostart/'; },
+  getFilePath(appName: string): string { return `${this.getDirectory()}${appName}.desktop`; }
+}
+
+const macAL = {
+  async enable({ appName, appPath, hidden }: AutoLaunchParams): Promise<ChildReturnData> {
+    const isHiddenValue = hidden ? 'true' : 'false';
+    const properties = `{path:"${appPath}", hidden:${isHiddenValue}, name:"${appName}"}`;
+    console.log('properties', properties);
+    return native.execApplescriptCommand(`make login item at end with properties ${properties}`);
+  },
+  async disable(appName: string): Promise<ChildReturnData> {
+    return native.execApplescriptCommand(`delete login item \"${appName}\"`);
+  },
+  async isEnabled(appName: string): Promise<boolean> {
+    const response: ChildReturnData = await native.execApplescriptCommand('get the name of every login item');
+    const loginList = response?.stdout[0]?.split(', ') || []
+    const exists = loginList.includes(appName)
+    console.log('login Item Exists:', exists);
+    return exists
+  },
+
+}
+
+const winAL = {
+  subKey: "SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Run",
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```

Comment:
Shouldn't you have TODO to support `hidden`? Also if I understand it, better name would be `minimized`.

Reply:
Minimized implies that there is a window which is minimized to the taskbar but what this actually means is don't spawn a widow at all. In hidden mode the app would just start in the tray on the respective platform with no window would be spawned until you open the tray menu. Added a TODO to handle this case on windows.

Reply:
What describe is that the window is effectively there, it is just not visible until you open the tray menu. I believe it is called `minimized` in most other apps, not `hidden`. In Linux at least there is actually a separate `hidden` attribute in the autostart file which will make it invisible in GUI that manages autostarting apps.

Reply:
Ok, I'm not sure of the specifics, this nomenclature came from the npm library I refactored which called it  isHiddenOnLaunch so I shortened to hidden. I can change to minimized.
https://github.com/Teamwork/node-auto-launch/blob/master/src/AutoLaunchWindows.coffee